Update for wlroots 0.18

Consolatis opened the pull request:
Same story as
- #626

This branch is not intended to be merged anytime soon but should allow to collaborate on chasing wlroots master.
As in the previous PR, feel free to push changes into this branch.

I suggest not using force pushes for now to reduce chances of accidentally overwriting something. If some existing commit should be amended just use `!fixup` or `[fixup]` or something along that line and refer to the subject of the commit to be amended (the hash may change due to rebases). We can remove that no-force-push restriction once we have a somewhat stable base to work with.

Like in #626 our goal should also be to make this PR as small as possible to reduce rebase conflicts with master. So if we can merge a commit in master that works for 0.17 and 0.18 without introducing any unnecessary complexity we should likely do that.

@ahesford @jlindgren90 @johanmalm 

Further information:
- breaking changes for 0.18: https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/3760
- #1260

Current issues:
- [ ] https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4600 seems to implement a immediate scale event for surfaces within scene. We may be able to remove our downstream fix (4238d7fc33dc2c7c4ee4ecbac81decbc91833c1a) for that issue, @ahesford could say more about that

Nice to have:
- [ ] listen to `wlr_renderer.events.lost` to catch GPU resets and do a `Reconfigure` to reload all compositor side buffers
- [ ] explicit sync ([wlroots MR](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4262), [sway PR](https://github.com/swaywm/sway/pull/8156))
- [ ] icon spec ([wlroots MR](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4704))
- [ ] adaptive_sync_supported ([wlroots MR](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4650))

Fixed:
- [x] ~~parentless xdg popups are not be handled correctly~~ (misunderstanding of the new redundant xdg-shell new_popup event)
- [x] the commit messages and `wlroots.wrap` files do not yet contain the correct wlroots hash and might also be out of order, I only did the first step of getting labwc to work with current wlroots master.
- [x] the initial xdg_surface configure flow has been changed, compositors now need to do that on their own
- [x] xdg decorations must not be confirmed directly but only if `xdg_surface->initialized == true`, also needs storage of the requested deco mode or link from view to the deco object itself which can be used from the xdg commit handler
- [x] at least when running nested the rounded corners sometimes flicker on mouse movements and are then completely transparent (caused by a wlroots bug, [MR pending](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4655))
- [x] as wlroots removed `wlr_output->pending` we have to use the `wlr_output_state_()` variants everywhere. But to reduce unnecessary screen blanking during mode changes we likely want to keep using a shared pending state. Rather than changing everything around to that I added a new `src/output-state.c` file which provides wrappers for the old `wlr_output_set_()`functions. This massively reduces rebase conflicts.
- [x] we currently crash on `ctrl-c` in gdb when running nested, ~~needs investigation~~ this is a libwayland bug, fixed via [this MR](https://gitlab.freedesktop.org/wayland/wayland/-/merge_requests/366) but not yet released
- [x] `lab_wlr_scene_commit()` still contains a comment about handling damage, that was an issue in wlroots 0.17.x but might now have been fixed upstream, needs to be verified
- [x] The CI is currently not set up to deal with wlroots compilations, should be easy to fix
- [x] `WLR_SCENE_DEBUG_DAMAGE=highlight` damage tracking is pretty broken, [wlroots MR](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4654)

There might be more issues, only slightly tested.

kode54 commented:
Crash on xdg popup from Steam's menus:

```
#0  __pthread_kill_implementation
    (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007f915eeab393 in __pthread_kill_internal (signo=6, threadid=<optimized out>)
    at pthread_kill.c:78
#2  0x00007f915ee5a6c8 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007f915ee424b8 in __GI_abort () at abort.c:79
#4  0x00007f915ee423dc in __assert_fail_base
    (fmt=0x7f915efbcb68 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x7f915ff29718 "!xsurface->override_redirect", file=file@entry=0x7f915ff29455 "xwayland/xwm.c", line=line@entry=1104, function=function@entry=0x7f915ff579e0 <__PRETTY_FUNCTION__.4.lto_priv.25> "wlr_xwayland_surface_restack") at assert.c:94
#5  0x00007f915ee52d46 in __assert_fail
    (assertion=assertion@entry=0x7f915ff29718 "!xsurface->override_redirect", file=file@entry=0x7f915ff29455 "xwayland/xwm.c", line=line@entry=1104, function=function@entry=0x7f915ff579e0 <__PRETTY_FUNCTION__.4.lto_priv.25> "wlr_xwayland_surface_restack") at assert.c:103
#6  0x00007f915ff10242 in wlr_xwayland_surface_restack
    (xsurface=xsurface@entry=0x5c3f5f243bf0, sibling=sibling@entry=0x0, mode=mode@entry=XCB_STACK_MODE_ABOVE) at ../wlroots-hidpi-xprop-git/xwayland/xwm.c:1104
#7  0x00005c3f5d1c015a in handle_map (listener=0x5c3f5f243f20, data=<optimized out>)
    at ../labwc/src/xwayland-unmanaged.c:45
#8  0x00007f915ff8d00e in wl_signal_emit_mutable (signal=<optimized out>, data=0x0)
    at ../wayland-1.22.0/src/wayland-server.c:2241
#9  0x00007f915ff8d00e in wl_signal_emit_mutable
    (signal=signal@entry=0x5c3f5eb95978, data=data@entry=0x5c3f5eb956a0)
    at ../wayland-1.22.0/src/wayland-server.c:2241
#10 0x00007f915feec1fd in surface_commit_state (surface=0x5c3f5eb956a0, next=0x5c3f5eb95820)
    at ../wlroots-hidpi-xprop-git/types/wlr_compositor.c:551
#11 0x00007f915f728596 in ffi_call_unix64 () at ../src/x86/unix64.S:104
#12 0x00007f915f72500e in ffi_call_int
    (cif=cif@entry=0x7ffd350a8600, fn=<optimized out>, rvalue=<optimized out>, avalue=<optimized out>, closure=closure@entry=0x0) at ../src/x86/ffi64.c:673
#13 0x00007f915f727bd3 in ffi_call
    (cif=cif@entry=0x7ffd350a8600, fn=<optimized out>, rvalue=rvalue@entry=0x0, avalue=avalue@entry=0x7ffd350a86d0) at ../src/x86/ffi64.c:710
#14 0x00007f915ff8bab0 in wl_closure_invoke
    (closure=closure@entry=0x5c3f5f2393a0, target=<optimized out>, 
    target@entry=0x5c3f5eddd7f0, opcode=opcode@entry=6, data=<optimized out>, 
    data@entry=0x5c3f5ea5a9f0, flags=2) at ../wayland-1.22.0/src/connection.c:1025
#15 0x00007f915ff90198 in wl_client_connection_data
    (fd=<optimized out>, mask=<optimized out>, data=<optimized out>)
    at ../wayland-1.22.0/src/wayland-server.c:438
#16 0x00007f915ff8eb02 in wl_event_loop_dispatch (loop=0x5c3f5dd64fd0, timeout=timeout@entry=-1)
    at ../wayland-1.22.0/src/event-loop.c:1027
#17 0x00007f915ff8f317 in wl_display_run (display=0x5c3f5dd64ee0)
    at ../wayland-1.22.0/src/wayland-server.c:1493
#18 0x00005c3f5d19a849 in main (argc=<optimized out>, argv=<optimized out>)
    at ../labwc/src/main.c:179
```

kode54 commented:
Further testing from earlier, shows that my screen doesn't repaint smoothly, either. Neither screen, in fact. And it has no effect whether adaptive sync is enabled or not. I cannot tell whether it's commit or damage related.

Consolatis commented:
The damage issues and wayland native windows not be appearing should hopefully be fixed by the last commit. Also rebased on top of master.

kode54 commented:
Further outstanding issues:

- [ ] The display refreshes at less than its native refresh rate. More like 20-30fps instead of 60. This doesn't seem to be related to use of adaptive sync on one of the displays.
- [ ] Atomic modesetting tearing support, which is supposed to be enabled upstream without any changes to the compositor to allow it, other than removing your check for the `WLR_DRM_NO_ATOMIC` variable, doesn't seem to work. I can try to pull DRM debug logs, but last I checked, it was wlroots committing things it shouldn't be committing during the atomic flips. Hell, AMD doesn't even want it to be committing cursor plane updates to allow tearing.

kode54 commented:
Logs for the above issues, possibly also relevant to wlroots 0.18 instead of just labwc:

[dmesg.log.gz](https://github.com/labwc/labwc/files/15124947/dmesg.log.gz)
[labwc.log.gz](https://github.com/labwc/labwc/files/15124949/labwc.log.gz)

kode54 commented:
Confirmed these frame rate issues do not occur with Sway master, so it's down to labwc and this PR being updated to handle whatever the refresh issue is.

Consolatis commented:
@kode54 added a FPS counter per output via debug log, maybe this helps to shed some light on your half monitor refresh rate observation. I have a slight feeling that this has something to do with the presentation time protocol but I am not sure. At least nested I get the full fps via `es2gears_x11` and `es2gears_wayland`. The FPS counter via debug log shows a reduction of the fps to about screen refresh rate in case of `es2gears_x11`, for the `_wayland` case the fps roughly matches between the es2gears output and labwc debug log.

johanmalm commented:
As discussed the other day, I've been running this PR for a couple of days now and it seems fine.

I've also carried out an independent review by studying the PR and the sway commit-history between 2023-11-21 and 2024-05-12 (i.e. since wlroots 0.17.0 release). See findings below.

## wlroots chasing - can be done on master now (I think)

- [ ] https://github.com/swaywm/sway/commit/bf2b79b2845a070d84aafaf95e6cfcf9af7eeb9b
- [ ] https://github.com/swaywm/sway/commit/dcdb72757a5ec591c692df5e96c57c51758dbd8f

## wlroots chasing required on `0.18` PR

There are three that I couldn't see covered in the PR.

- [x] https://github.com/swaywm/sway/commit/2058209a130f5051b59d8ebb24196409695deaaf
- [x] https://github.com/swaywm/sway/commit/fca8474e9bd64bff8df16fdaf409d5f575ba9501
- [x] https://github.com/swaywm/sway/commit/f202bc84d2c072ae39fc4e8a350a8b0a616f350b
- [x] https://github.com/swaywm/sway/commit/a4e85332a1fe6b079c6da1689ca33d8ba4c71977
- [ ] https://github.com/swaywm/sway/commit/22d0dd8bde62592ff34d354e059cdc6e09879cb4
- [ ] https://github.com/swaywm/sway/commit/f12023b1a26e30b7c041551a4713ab81014d138d
- [ ] https://github.com/swaywm/sway/commit/47e6a1164c25b5b49bfb919a681b88641d2ce37c
- [x] https://github.com/swaywm/sway/commit/128b6253a924c30e8e4fd1f09ce19e21a826c5ad
- [x] https://github.com/swaywm/sway/commit/a946b1aecfa3f32a3bfae66d66df0bdf77440d69

## wlroots chasing not required by labwc

- https://github.com/swaywm/sway/commit/9e1465107788af2c8ce93e2a288e9d32bc09711c
- https://github.com/swaywm/sway/commit/ffcde7a70c1341d1ec1d38c00ff87faa1a816892
- https://github.com/swaywm/sway/commit/7e69a7076fc8a4eb788e0229b1c99dd0b7b04bb7
- https://github.com/swaywm/sway/commit/64d644f0da15e4aef85ca1800b90aa65e333933c
- https://github.com/swaywm/sway/commit/fd6d6f1d97deea2fc94aaeb58c85c99a498d905b

## Things in `0.18 PR` that may need more work

### xdg-shell (toplevels and popups)

Ref: https://github.com/labwc/labwc/pull/1641/commits/cb8d89e9563f8521883a60d956b09c2f2ee2b86e

It looks like there is a lot of xdg/layer-shell toplevel/popup stuff that has changed. Found the following commits that I don't think we've not fully covered in PR 0.18. Guess some of them can be done on master now.

- [ ] 1. https://github.com/swaywm/sway/commit/47e6a1164c25b5b49bfb919a681b88641d2ce37c
- [ ] 2. https://github.com/swaywm/sway/commit/3ef5abd405a6fd32aeeffb2f48a6cadd9fc14574
- [ ] 3. https://github.com/swaywm/sway/commit/4e6d7612ffbd8e29713ae063937c8460e091bb75
- [ ] 4. https://github.com/swaywm/sway/commit/1846944f0454fec08ad91a04ab04e3b1ffcd1764
- [ ] 5. https://github.com/swaywm/sway/commit/09c360d503eb7309de46af3745001df7dd911438
- [ ] 6. https://github.com/swaywm/sway/commit/1dc661af176d80db8b283e9c38fd127f71d33052

### XWayland re-stacking

I'm sure you're right on this, but couldn't find anything in sways commit log that gave me confidence that we're approaching it right. How do we gain confidence that this XWayland restacking stuff is right? What changed in wlroots to trigger this amendment?

- [ ] https://github.com/labwc/labwc/pull/1641/commits/0f903ccf23feabbc0fd2d9181879ba614bf70e7f

# Aside

As an aside, I spotted the following on sway `master` and think we could consider these at some point (but not pre-req to `0.18` merge). We could probably start a separate issue with these later, but am including them here for the minute for completeness.

- [ ] a. https://github.com/swaywm/sway/commit/2b08e79061db583d0ffe46114c48f958a0d3e6ed
- [ ] b. https://github.com/swaywm/sway/commit/2e951163c5a5f24fe9cf7ee348e56b09719a99a9
- [ ] c. https://github.com/swaywm/sway/commit/f2a0e81b2438853e12a2b8fe9bddde154852d85d
- [ ] d. https://github.com/swaywm/sway/commit/59f629238309e230b0e353e73d4f37a7de7fe820
- [ ] e. https://github.com/swaywm/sway/commit/fc640d5f6c82883c35e90a64f0098486e6091293
- [ ] f. https://github.com/swaywm/sway/commit/d19810eba8959f052d91fd6609cef6adf36b3951
- [ ] g. https://github.com/swaywm/sway/commit/1b5515400d0e10c8e1205b88cfc0e46ecb5faa96
- [ ] h. https://github.com/swaywm/sway/commit/c5fd8c050f7ddbfe3e5b7abc8f5f6ace3a3c5307
- [ ] i. https://github.com/swaywm/sway/commit/a6ef12d968111d4d8c480124575d9d86ae61e470
- [ ] j. https://github.com/swaywm/sway/commit/ff07eab85b5c6b728ad3cc99d02e9f2ae8b9854f

Consolatis commented:
> I'm sure you're right on this, but couldn't find anything in sways commit log that gave me confidence that we're approaching it right. How do we gain confidence that this XWayland restacking stuff is right? What changed in wlroots to trigger this amendment?

wlroots now handles that internally (and in fact asserts that the given surface is not an unamanged one), see https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4052. That was implemented in wlroots to fix a sway bug because it did not restack unmanged surfaces on top like we do / did.

kode54 commented:
Frame rate issues seem to be resolved now.

Consolatis commented:
Rebased on master and disabled the fps counter, can be re-enabled by uncommenting `#define FPS_TIMEOUT (2 * 1000)` in `src/output-state.c`.

kode54 commented:
Closing virt-manager window for a VM now crashes with memory corruption:

```
#0  __pthread_kill_implementation
    (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0)
    at pthread_kill.c:44
#1  0x00007604079e4eb3 in __pthread_kill_internal
    (threadid=<optimized out>, signo=6) at pthread_kill.c:78
#2  0x000076040798ca30 in __GI_raise (sig=sig@entry=6)
    at ../sysdeps/posix/raise.c:26
#3  0x00007604079744c3 in __GI_abort () at abort.c:79
#4  0x0000760407975354 in __libc_message_impl
    (fmt=fmt@entry=0x760407afe2ea "%s\n") at ../sysdeps/posix/libc_fatal.c:132
#5  0x00007604079ef085 in malloc_printerr
    (str=str@entry=0x760407afbffe "corrupted double-linked list")
    at malloc.c:5772
#6  0x00007604079efbfc in unlink_chunk
    (p=p@entry=0x5af189f7bcf0, av=0x760407b32ac0 <main_arena>) at malloc.c:1617
#7  0x00007604079f2a9a in _int_malloc
    (av=av@entry=0x760407b32ac0 <main_arena>, bytes=bytes@entry=208)
    at malloc.c:4381
#8  0x00007604079f477e in __libc_calloc
    (n=n@entry=1, elem_size=elem_size@entry=208) at malloc.c:3754
#9  0x0000760408397130 in zalloc (s=208)
    at ../wayland-1.22.0/src/wayland-private.h:234
#10 wl_closure_init
    (message=message@entry=0x760408385fb0 <zwlr_foreign_toplevel_handle_v1_events.lto_priv+144>, size=size@entry=0, num_arrays=num_arrays@entry=0x0, args=args@entry=0x7ffe3d108350) at ../wayland-1.22.0/src/connection.c:572
#11 0x0000760408399abf in wl_closure_marshal
    (message=0x760408385fb0 <zwlr_foreign_toplevel_handle_v1_events.lto_priv+144>, args=0x7ffe3d108350, opcode=6, sender=0x5af189f9c7a0)
    at ../wayland-1.22.0/src/connection.c:608
#12 handle_array
    (resource=resource@entry=0x5af189f9c7a0, opcode=opcode@entry=6, args=args@entry=0x7ffe3d108350, send_func=send_func@entry=0x7604083976f0 <wl_closure_send>)
    at ../wayland-1.22.0/src/wayland-server.c:224
#13 0x0000760408399cf0 in wl_resource_post_event_array
    (resource=resource@entry=0x5af189f9c7a0, opcode=opcode@entry=6, args=args@entry=0x7ffe3d108350) at ../wayland-1.22.0/src/wayland-server.c:244
#14 0x0000760408399ddf in wl_resource_post_event
    (resource=resource@entry=0x5af189f9c7a0, opcode=opcode@entry=6)
    at ../wayland-1.22.0/src/wayland-server.c:259
#15 0x00007604082f3eb8 in zwlr_foreign_toplevel_handle_v1_send_closed
    (resource_=0x5af189f9c7a0)
    at protocol/wlr-foreign-toplevel-management-unstable-v1-protocol.h:490
#16 wlr_foreign_toplevel_handle_v1_destroy (toplevel=0x5af189f9c6c0)
    at ../wlroots-hidpi-xprop-git/types/wlr_foreign_toplevel_management_v1.c:499
#17 wlr_foreign_toplevel_handle_v1_destroy (toplevel=0x5af189f9c6c0)
    at ../wlroots-hidpi-xprop-git/types/wlr_foreign_toplevel_management_v1.c:489
#18 0x00005af17d759615 in view_destroy (view=0x5af189f7be60)
    at ../labwc/src/view.c:2325
#19 0x000076040839900e in wl_signal_emit_mutable
    (signal=signal@entry=0x5af189ef4da8, data=data@entry=0x0)
    at ../wayland-1.22.0/src/wayland-server.c:2241
#20 0x00007604082ec0c8 in destroy_xdg_surface (surface=0x5af189ef4d00)
    at ../wlroots-hidpi-xprop-git/types/xdg_shell/wlr_xdg_surface.c:492
#21 0x00007604082f1e0d in surface_destroy_role_object (surface=0x5af189ef48a0)
    at ../wlroots-hidpi-xprop-git/types/wlr_compositor.c:910
#22 surface_destroy_role_object (surface=0x5af189ef48a0)
    at ../wlroots-hidpi-xprop-git/types/wlr_compositor.c:904
#23 surface_handle_role_resource_destroy
    (listener=0x5af189ef4bf0, data=<optimized out>)
    at ../wlroots-hidpi-xprop-git/types/wlr_compositor.c:891
#24 0x000076040839a99f in wl_priv_signal_final_emit
    (data=<optimized out>, signal=<optimized out>)
    at ../wayland-1.22.0/src/wayland-server.c:2377
#25 destroy_resource (element=0x5af189ef4e50, data=data@entry=0x0, flags=0)
    at ../wayland-1.22.0/src/wayland-server.c:729
#26 0x000076040839c7e9 in wl_resource_destroy (resource=<optimized out>)
    at ../wayland-1.22.0/src/wayland-server.c:749
#27 0x00007604073fc596 in ffi_call_unix64 () at ../src/x86/unix64.S:104
#28 0x00007604073f900e in ffi_call_int
    (cif=cif@entry=0x7ffe3d1087e0, fn=<optimized out>, rvalue=<optimized out>, avalue=<optimized out>, closure=closure@entry=0x0) at ../src/x86/ffi64.c:673
#29 0x00007604073fbbd3 in ffi_call
    (cif=cif@entry=0x7ffe3d1087e0, fn=<optimized out>, rvalue=rvalue@entry=0x0, avalue=avalue@entry=0x7ffe3d1088b0) at ../src/x86/ffi64.c:710
#30 0x0000760408397ab0 in wl_closure_invoke
    (closure=closure@entry=0x5af189ed0500, target=<optimized out>,
    target@entry=0x5af189ef4e50, opcode=opcode@entry=0, data=<optimized out>,
    data@entry=0x5af189e79b40, flags=2)
    at ../wayland-1.22.0/src/connection.c:1025
#31 0x000076040839c198 in wl_client_connection_data
    (fd=<optimized out>, mask=<optimized out>, data=<optimized out>)
    at ../wayland-1.22.0/src/wayland-server.c:438
#32 0x000076040839ab02 in wl_event_loop_dispatch
    (loop=0x5af188923c20, timeout=timeout@entry=-1)
    at ../wayland-1.22.0/src/event-loop.c:1027
#33 0x000076040839b317 in wl_display_run (display=0x5af188923b30)
    at ../wayland-1.22.0/src/wayland-server.c:1493
#34 0x00005af17d7436e3 in main (argc=<optimized out>, argv=<optimized out>)
    at ../labwc/src/main.c:217
```

kode54 commented:
Updated wlroots, one new commit dealing with direct scanout and color processing, and now the above crash isn't happening any longer?

kode54 commented:
This apparently fixes the quarter screen issue I was having with Waydroid? Seems to work without requiring me to suspend my displays first.

droc12345 commented:
I've been running this version for a while, and turned on logging this morning and noticed error messages

`00:00:00.182 [ERROR] [types/wlr_layer_shell_v1.c:296] A configure is sent to an uninitialized wlr_layer_surface_v1 0x17d9a80`

It doesn't seem to matter which window, every time I open a new app/window, the message appears.
conky, foot, etc trigger it, whatever it is I start, doesn't seem to affect the running.
I've updated everything this morning to check, wlroots, foot, and several apps, still the messages.

Consolatis commented:
> The error messages come from wlr_xdg_surface_schedule_configure, and it's only called in the
> one place, but I only see the "scheduling configure" message occasionally, So I'm not sure what
> exactly is triggering off the messages.

Its caused by the decoration manager:
> xdg decorations must not be confirmed directly but only if xdg_surface->initialized == true, also needs storage of the requested deco mode or link from view to the deco object itself which can be used from the xdg commit handler

droc12345 commented:
It does report both depending on whether a layer app or regular app.

kode54 commented:
Is this worth rebasing without addressing the issues mentioned above? Or is there some pending work already?

Consolatis commented:
> Is this worth rebasing without addressing the issues mentioned above?

Yes, rebased on current master.

> Or is there some pending work already?

No, didn't yet look into them.

kode54 commented:
I'll rebuild again tonight, but there was a bug I noticed with Fullscreen toggle, if toggling from a Maximized window. If toggling from Maximized, it wants to resize the window, but fit it within the same position offset as Maximized, and leaves the dock overlay onscreen.

kode54 commented:
Is this worth chasing again so close to a new release window? I'm tracking wlroots 0.17.4 as well.

droc12345 commented:
I'm using this daily, along with the view cascade patch and my fixes for window cycling and it's working well.
I compiled yesterday, using the latest wlroots and no problems there.

droc12345 commented:
Latest wlroots, had to patch for a protocol name change, due to going from unstable to stable.

`-   wl_protocol_dir / 'unstable/tablet/tablet-unstable-v2.xml',`
`+   wl_protocol_dir / 'stable/tablet/tablet-v2.xml',`

Consolatis commented:
> Latest wlroots, had to patch for a protocol name change, due to going from unstable to stable.
> 
> `- wl_protocol_dir / 'unstable/tablet/tablet-unstable-v2.xml',` `+ wl_protocol_dir / 'stable/tablet/tablet-v2.xml',`

Thanks. Rebased, changed to the stable variant + raised the w-p version required to 1.35.

droc12345 commented:
Couple of wlr_output->pending need to be changed as pending has gone from struct wlr_output.
one in scene-helpers.c and the other in output.c
I changed them to output->pending on my end, compiles ok, not sure about whether it works, as that's in
the gamma code and I don't use that.

kode54 commented:
labwc on this branch patch, applied to master as of 0.7.2.r96.g370d62be, now consistently locks up my local console upon logging out. I cannot reproduce with sway-git on the same wlroots, which is 0.18.0.20240630.060115-1.

The lock up is weird. getty doesn't respond to keyboard input, and capslock light doesn't work, nor do ctrl+fx or ctrl+alt+fx. But logging in via SSH, I can poll the keyboard with evtest successfully and receive input events from keypresses.

droc12345 commented:
Not seeing any lockups on my end. I just compiled and tested. (using latest labwc and wlroots)

All the 370d62be patch did was remove the seatd.wrap, the commit before deals with ssd (I run with no decorations) 

Seatd shouldn't cause any problems unless you were using the wrap file to install it.
If so, then use the normal seatd package (whatever your distro is)

kode54 commented:
Just using these two packages in a clean chroot, using `aur -S --chroot` (from `aurutils`)

https://aur.archlinux.org/packages/wlroots-hidpi-xprop-git
https://aur.archlinux.org/packages/labwc-wlroots-018-git

The only difference is that instead of building this branch as-is, I am taking its patch series and applying them to the master branch.

kode54 commented:
Disregard my report. It was a package called `airstatus` that was leaking dbus resources, and it wrecked my dbus broker.

droc12345 commented:
Looks like they've released 0.18.0, need to change **master** to **0.18.0** in _wlroots.wrap_

droc12345 commented on subprojects/wlroots.wrap:
master now points to 0.19-dev

Consolatis commented:
Debian and Void currently miss libdrm >= 2.4.122 thus they fail in the CI. Decided against putting a libdrm.wrap file in to not surprise users.

Consolatis commented:
Decorations should now be handled correctly thus no more "xdg configure before surface initialized" errors.

Commits now update the `wlroots.wrap` revision and are sorted by the wlroots commit history so `git bisect` and friends are happy. Branch still needs minor work (commit subjects, some need to be squashed) but the vast majority within the scope of this PR should be done.

yuiiio commented:
wlroots 0.18 and this branch seems solve
weston-simple-dmabuf-feedback: assert
```
../clients/simple-dmabuf-feedback.c:573: create_dmabuf_buffer: Assertion `buf->bo && "error: could not create GBM bo for buffer"' failed.
```
mentioned at https://github.com/labwc/labwc/pull/1278#issuecomment-1868507772

yuiiio commented:
I noticed a difference in behavior about mpv fullscreen & VRR between (master + dmabuf-feedback patch) and (0.18 branch + dmabuf-feedback patch)

* master(wlroots 0.17):
        mpv (fullscreen/windowed) VRR works with/without dmabuf-feedback patch. (both --vo=gpu/--vo=dmabuf-wayland)
* (wlroots-0.18 branch): 
        mpv fullscreen VRR doesn't work(fixed to max rate) without dmabuf-feedback patch & --vo=dmabuf-wayland
        mpv windowed VRR works without dmabuf-feedback patch.
        WLR_SCENE_DISABLE_DIRECT_SCANOUT=1 allows to work vrr again without dmabuf-feedback patch & --vo=dmabuf-wayland.

not sure if it's a bug or not in wlroots.
some tested Xwayland games seems doesn't affect this.

kode54 commented:
> mpv

@yuiiio I can't even use VRR right now, because my GPU doesn't like resetting and restoring VRR functionality properly. Following a GPU recovery event, or following a DPMS power cycle of my VRR monitor, I need to restart labwc, or else it will blank the screen when it means to VRR. I need to figure something out. I have reported the issue to drm/amd as well.

droc12345 commented:
wlroots 0.18 has some work for gpu recovery, but not sure to what extent.
But anything dealing with gpu has to take into account, the video card, libdrm, amd firmware and kernel drivers.

I have rx550/560 and they suffer from a reset bug that will never be fixed, at least from amd and the kernel.
But I've since replaced them with newer cards.
Although if you go too new, then you run into the problems of firmware not being available or the kernel driver not
up to date.

ahesford commented:
Regarding the scale-notification fix, I'll have to test, but I suspect that, at best, we might be able to eliminate [our call to `wlr_fractional_scale_v1_notify_scale`](https://github.com/labwc/labwc/blob/master/src/xdg.c#L777-L778). My intuition is that we should keep the rest of the change in https://github.com/labwc/labwc/commit/4238d7fc33dc2c7c4ee4ecbac81decbc91833c1a (namely, assigning an output in `xdg_surface_new` before the view is mapped).

johanmalm commented:
> I think this is ready to go. Thanks everybody for testing and reporting bugs.

Consolatis. Big thanks. You're a star. Looks super neat.

Consolatis commented:
@Narrat the master branch / `labwc-git` is now on wlroots 0.18.

Narrat commented:
Great :) [Pushed](https://aur.archlinux.org/cgit/aur.git/commit/?h=labwc-git&id=46ca1523e573d16e44065acbb17f0e20242c76ec) the adjustment to wlroots 0.18
